File: frontend/src/types/qsar_ready.ts
/**
 * TypeScript interfaces for the QSAR-Ready Pipeline API.
 * Covers all endpoints under /api/v1/qsar-ready/.
 *
 * All shapes match backend Pydantic schemas in backend/app/schemas/qsar_ready.py.
 */

// ====================== Config ======================

/**
 * Configuration for the QSAR-Ready curation pipeline.
 * Controls which steps are enabled and molecular property filter thresholds.
 */
export interface QSARReadyConfig {
  enable_metals: boolean;
  enable_desalt: boolean;
  enable_normalize: boolean;
  enable_neutralize: boolean;
  enable_tautomer: boolean;
  enable_stereo_strip: boolean;
  enable_isotope_strip: boolean;
  min_heavy_atoms: number;
  max_heavy_atoms: number;
  max_mw: number;
  remove_inorganics: boolean;
}

// ====================== Single Molecule Result ======================

/**
 * Result for a single pipeline step.
 * Tracks SMILES before/after the step for provenance.
 */
export interface QSARStepResult {
  step_name: string;
  step_index: number;
  enabled: boolean;
  status: 'applied' | 'no_change' | 'skipped' | 'error';
  before_smiles: string | null;
  after_smiles: string | null;
  detail: string | null;
}

/**
 * Full curation result for a single molecule.
 * Carries both original and standardized InChIKeys to track identity changes.
 */
export interface QSARReadyResult {
  original_smiles: string;
  original_inchikey: string | null;
  curated_smiles: string | null;
  standardized_inchikey: string | null;
  inchikey_changed: boolean;
  status: 'ok' | 'rejected' | 'duplicate' | 'error';
  rejection_reason: string | null;
  steps: QSARStepResult[];
}

// ====================== Batch ======================

/**
 * Summary statistics for a completed batch QSAR-ready job.
 */
export interface QSARBatchSummary {
  total: number;
  ok: number;
  rejected: number;
  duplicate: number;
  error: number;
  steps_applied_counts: Record<string, number>;
}

/**
 * Response from the batch upload endpoint.
 * Returns job_id for polling status and fetching results.
 */
export interface QSARBatchUploadResponse {
  job_id: string;
  total_molecules: number;
  status: string;
  message: string;
}

/**
 * Real-time status response for a batch job (WebSocket + polling).
 */
export interface QSARBatchStatusResponse {
  job_id: string;
  status: 'pending' | 'processing' | 'complete' | 'failed';
  progress: number;
  processed: number;
  total: number;
}

/**
 * Paginated results response for a completed batch job.
 */
export interface QSARBatchResultsResponse {
  job_id: string;
  status: string;
  config: QSARReadyConfig;
  summary: QSARBatchSummary;
  results: QSARReadyResult[];
  page: number;
  per_page: number;
  total_pages: number;
}

// ====================== Pipeline Step Metadata ======================

/**
 * Metadata for a single configurable pipeline step.
 * Used to render the step toggle grid in the UI.
 */
export interface QSARPipelineStep {
  key: keyof Omit<QSARReadyConfig, 'min_heavy_atoms' | 'max_heavy_atoms' | 'max_mw' | 'remove_inorganics'>;
  index: number;
  name: string;
  description: string;
}

/**
 * Saved pipeline config profile for localStorage persistence.
 */
export interface QSARPipelineProfile {
  id: string;
  name: string;
  config: QSARReadyConfig;
  createdAt: number;
}

// ====================== Constants ======================

/** Canonical step order for UI rendering. Mirrors backend 10-step pipeline. */
export const PIPELINE_STEPS: QSARPipelineStep[] = [
  { key: 'enable_metals', index: 2, name: 'Metal Disconnect', description: 'Remove metal bonds and ions' },
  { key: 'enable_desalt', index: 3, name: 'Desalt', description: 'Keep largest organic fragment' },
  { key: 'enable_normalize', index: 4, name: 'Normalize', description: 'Fix functional group representations' },
  { key: 'enable_neutralize', index: 5, name: 'Neutralize', description: 'Remove charges where possible' },
  { key: 'enable_tautomer', index: 6, name: 'Tautomer', description: 'Canonicalize tautomeric form' },
  { key: 'enable_stereo_strip', index: 7, name: 'Strip Stereo', description: 'Remove stereochemistry information' },
  { key: 'enable_isotope_strip', index: 8, name: 'Strip Isotopes', description: 'Remove isotope labels' },
];

/** Default configs for the three built-in presets. */
export const PRESET_CONFIGS: Record<string, QSARReadyConfig> = {
  'qsar-2d': {
    enable_metals: true,
    enable_desalt: true,
    enable_normalize: true,
    enable_neutralize: true,
    enable_tautomer: true,
    enable_stereo_strip: true,
    enable_isotope_strip: true,
    min_heavy_atoms: 3,
    max_heavy_atoms: 100,
    max_mw: 1500.0,
    remove_inorganics: true,
  },
  'qsar-3d': {
    enable_metals: true,
    enable_desalt: true,
    enable_normalize: true,
    enable_neutralize: true,
    enable_tautomer: true,
    enable_stereo_strip: false,
    enable_isotope_strip: true,
    min_heavy_atoms: 3,
    max_heavy_atoms: 100,
    max_mw: 1500.0,
    remove_inorganics: true,
  },
  'minimal': {
    enable_metals: false,
    enable_desalt: true,
    enable_normalize: true,
    enable_neutralize: false,
    enable_tautomer: false,
    enable_stereo_strip: false,
    enable_isotope_strip: false,
    min_heavy_atoms: 3,
    max_heavy_atoms: 100,
    max_mw: 1500.0,
    remove_inorganics: false,
  },
};

File: backend/app/schemas/qsar_ready.py
"""
Pydantic v2 schemas for QSAR-Ready Pipeline (Phase 10).

Provides request and response models for the QSAR-Ready API surface:
- Single molecule processing
- Batch file upload and status
- Batch results with pagination and summary statistics
"""

from typing import Optional

from pydantic import BaseModel, Field


# =============================================================================
# Config Schema
# =============================================================================


class QSARReadyConfigSchema(BaseModel):
    """Pipeline configuration for QSAR-Ready curation.

    Mirrors QSARReadyConfig dataclass from pipeline.py.
    All boolean toggle fields default to match the default QSARReadyConfig.
    """

    model_config = {"from_attributes": True}

    enable_metals: bool = Field(
        default=True,
        description="Disconnect metal-organic bonds (step 2: MetalDisconnector)",
    )
    enable_desalt: bool = Field(
        default=True,
        description="Keep largest fragment / remove salts (step 3: LargestFragmentChooser)",
    )
    enable_normalize: bool = Field(
        default=True,
        description="Normalize functional groups (step 4: Normalizer)",
    )
    enable_neutralize: bool = Field(
        default=True,
        description="Neutralize charges (step 5: Uncharger)",
    )
    enable_tautomer: bool = Field(
        default=True,
        description="Canonicalize tautomers (step 6: TautomerEnumerator)",
    )
    enable_stereo_strip: bool = Field(
        default=False,
        description="Strip stereochemistry (step 7). True for QSAR-2D, False for QSAR-3D.",
    )
    enable_isotope_strip: bool = Field(
        default=True,
        description="Strip isotope labels (step 8).",
    )
    min_heavy_atoms: int = Field(
        default=3,
        ge=1,
        le=500,
        description="Minimum heavy atom count for composition filter (step 9)",
    )
    max_heavy_atoms: int = Field(
        default=100,
        ge=1,
        le=500,
        description="Maximum heavy atom count for composition filter (step 9)",
    )
    max_mw: float = Field(
        default=1500.0,
        gt=0,
        le=5000.0,
        description="Maximum molecular weight (Da) for composition filter (step 9)",
    )
    remove_inorganics: bool = Field(
        default=True,
        description="Reject molecules with no carbon atoms in composition filter (step 9)",
    )


# =============================================================================
# Step and Result Schemas
# =============================================================================


class QSARStepResultSchema(BaseModel):
    """Result of a single QSAR-Ready pipeline step.

    Mirrors QSARStepResult dataclass from pipeline.py.
    """

    model_config = {"from_attributes": True}

    step_name: str = Field(
        description=(
            "Step identifier: parse, metals, desalt, normalize, neutralize, "
            "tautomer, stereo, isotope, filter, or canonical"
        )
    )
    step_index: int = Field(description="1-based step index (1 through 10)")
    enabled: bool = Field(description="Whether this step was enabled in the config")
    status: str = Field(
        description="Step outcome: 'applied', 'no_change', 'skipped', or 'error'"
    )
    before_smiles: Optional[str] = Field(
        default=None,
        description="SMILES before this step was applied (None for skipped steps)",
    )
    after_smiles: Optional[str] = Field(
        default=None,
        description="SMILES after this step was applied (None for skipped/error steps)",
    )
    detail: Optional[str] = Field(
        default=None,
        description="Human-readable detail: change description, rejection reason, or error message",
    )


class QSARReadyResultSchema(BaseModel):
    """Full result of processing a single molecule through the QSAR-Ready pipeline.

    Mirrors QSARReadyResult dataclass from pipeline.py.
    """

    model_config = {"from_attributes": True}

    original_smiles: str = Field(description="The input SMILES string")
    original_inchikey: Optional[str] = Field(
        default=None,
        description="InChIKey computed from parsed mol BEFORE any pipeline steps (D-14)",
    )
    curated_smiles: Optional[str] = Field(
        default=None,
        description="Canonical SMILES after all pipeline steps (None if rejected or error)",
    )
    standardized_inchikey: Optional[str] = Field(
        default=None,
        description="InChIKey computed from final mol AFTER all pipeline steps (D-14)",
    )
    inchikey_changed: bool = Field(
        default=False,
        description="True if original_inchikey != standardized_inchikey (D-14 locked decision)",
    )
    status: str = Field(
        description="Pipeline outcome: 'ok', 'rejected', 'duplicate', or 'error'"
    )
    rejection_reason: Optional[str] = Field(
        default=None,
        description="Human-readable reason for rejection or duplication",
    )
    steps: list[QSARStepResultSchema] = Field(
        default_factory=list,
        description="Per-step provenance list (10 entries when status != 'error')",
    )


# =============================================================================
# Request Models
# =============================================================================


class QSARSingleRequest(BaseModel):
    """Request body for single-molecule QSAR-Ready processing."""

    model_config = {"from_attributes": True}

    smiles: str = Field(
        ...,
        min_length=1,
        max_length=5000,
        description="SMILES string to process through the QSAR-Ready pipeline",
    )
    config: QSARReadyConfigSchema = Field(
        default_factory=QSARReadyConfigSchema,
        description="Pipeline configuration (defaults to standard QSAR-Ready config)",
    )


# =============================================================================
# Batch Response Models
# =============================================================================


class QSARBatchUploadResponse(BaseModel):
    """Response from submitting a QSAR-Ready batch job."""

    model_config = {"from_attributes": True}

    job_id: str = Field(description="Unique job identifier for tracking progress")
    total_molecules: int = Field(
        description="Number of molecules detected in the uploaded file"
    )
    status: str = Field(default="pending", description="Initial job status")
    message: str = Field(
        default="Job submitted successfully",
        description="Human-readable status message",
    )


class QSARBatchStatusResponse(BaseModel):
    """Current status of a QSAR-Ready batch job."""

    model_config = {"from_attributes": True}

    job_id: str = Field(description="Job identifier")
    status: str = Field(
        description="Job status: 'pending', 'processing', 'complete', or 'failed'"
    )
    progress: int = Field(default=0, description="Progress percentage (0-100)")
    processed: int = Field(default=0, description="Number of molecules processed so far")
    total: int = Field(default=0, description="Total number of molecules in the job")
    eta_seconds: Optional[int] = Field(
        default=None, description="Estimated seconds remaining until job completion"
    )


class QSARBatchSummary(BaseModel):
    """Aggregate statistics for a completed QSAR-Ready batch job."""

    model_config = {"from_attributes": True}

    total: int = Field(description="Total number of input molecules")
    ok: int = Field(description="Molecules successfully curated (status='ok')")
    rejected: int = Field(description="Molecules rejected by pipeline steps (status='rejected')")
    duplicate: int = Field(
        description="Molecules that are InChIKey duplicates of earlier molecules (status='duplicate')"
    )
    error: int = Field(description="Molecules that encountered pipeline errors (status='error')")
    steps_applied_counts: dict[str, int] = Field(
        default_factory=dict,
        description="Per-step count of molecules where the step was applied (status='applied')",
    )


class QSARBatchResultsResponse(BaseModel):
    """Paginated results response for a completed QSAR-Ready batch job."""

    model_config = {"from_attributes": True}

    job_id: str = Field(description="Job identifier")
    status: str = Field(
        description="Job status: 'pending', 'processing', 'complete', or 'failed'"
    )
    config: QSARReadyConfigSchema = Field(
        description="Pipeline configuration used for this batch job"
    )
    summary: QSARBatchSummary = Field(
        description="Aggregate statistics for the completed batch job"
    )
    results: list[QSARReadyResultSchema] = Field(
        default_factory=list,
        description="Paginated list of per-molecule pipeline results",
    )
    page: int = Field(default=1, description="Current page number (1-indexed)")
    per_page: int = Field(default=50, description="Number of results per page")
    total_pages: int = Field(default=1, description="Total number of result pages")
    total_results: int = Field(
        default=0, description="Total number of result records across all pages"
    )
